### PR TITLE
WT-10019 Add read/write random number generator support for test apps

### PR DESCRIPTION
**Pull request by ddanderson**

### CODE REVIEW

File: test/csuite/wt9937_parse_opts/main.c [hunk 1]
```diff
@@ -264,13 +289,26 @@ verify_expect(
...
+    } while (0)
 
     /*
      * opts->home is always set, even without -h on the command line, so don't check it here.
      */
     VERIFY_STR(opts, expect, build_dir);
     VERIFY_STR(opts, expect, tiered_storage_source);
     VERIFY_INT(opts, expect, table_type);
+    VERIFY_RANDOM_INT(opts, expect, write_seed);
+    VERIFY_RANDOM_INT(opts, expect, read_seed);
```
Comment (sueloverso) [inline test/csuite/wt9937_parse_opts/main.c:311]: Alphabetize.

File: test/csuite/wt9937_parse_opts/main.c [hunk 2]
```diff
@@ -355,9 +393,12 @@ main(int argc, char *argv[])
             expect.tiered_storage = subset->tiered_storage;
             expect.verbose = subset->verbose;
             expect.nthreads = subset->nthreads;
+            expect.write_seed = subset->write_seed;
+            expect.read_seed = subset->read_seed;
```
Comment (sueloverso) [inline test/csuite/wt9937_parse_opts/main.c:397]: Alphabetize.

File: test/utility/parse_opts.c [hunk 3]
```diff
@@ -47,10 +47,73 @@ extern int __wt_optreset;
...
+            seed_end = &seed_str[strlen(seed_str)];
+        parse_end = (char *)seed_end;
+        if (*seed_str == 'W') {
+            if (opts->write_seed != 0)
+                testutil_die(EINVAL, "-PS Wxxx write seed specified twice: %s", SEED_USAGE);
+            parse_seed(&opts->write_seed, seed_str + 1, &parse_end);
+        } else if (*seed_str == 'R') {
+            if (opts->read_seed != 0)
+                testutil_die(EINVAL, "-PS Rxxx read seed specified twice: %s, SEED_USAGE");
+            parse_seed(&opts->read_seed, seed_str + 1, &parse_end);
```
Comment (sueloverso) [inline test/utility/parse_opts.c:98]: I'd prefer to process these in alphabetical order too.

File: test/utility/parse_opts.c [hunk 4]
```diff
@@ -47,10 +47,73 @@ extern int __wt_optreset;
...
+ * parse_tiered_random_seeds --
  *     Parse a command line option for the tiered storage configurations.
  */
 static void
+parse_tiered_random_seeds(TEST_OPTS *opts, const char *seed_str)
+{
+    char *parse_end;
+    const char *seed_end;
+    static const char *SEED_USAGE =
+      "-PS parameter is comma separated write and read seeds, e.g. -PSW1234,R5678";
```
Comment (sueloverso) [inline test/utility/parse_opts.c:84]: Can these be talked about and presented in alphabetical order? "...separated read and write seed, e.g. -PSR1234,W5678"?

File: test/utility/parse_opts.c [hunk 5]
```diff
@@ -47,10 +47,73 @@ extern int __wt_optreset;
...
+            parse_seed(&opts->write_seed, seed_str + 1, &parse_end);
+        } else if (*seed_str == 'R') {
+            if (opts->read_seed != 0)
+                testutil_die(EINVAL, "-PS Rxxx read seed specified twice: %s, SEED_USAGE");
+            parse_seed(&opts->read_seed, seed_str + 1, &parse_end);
+        } else
+            testutil_die(EINVAL, "%s", SEED_USAGE);
+
+        if (*parse_end != '\0' && *parse_end != ',')
+            testutil_die(EINVAL, "-PS value after 'W' or 'R' is not an integer: %s", SEED_USAGE);
```
Comment (sueloverso) [inline test/utility/parse_opts.c:103]: Same here.

File: test/utility/parse_opts.c [hunk 6]
```diff
@@ -79,6 +148,8 @@ testutil_parse_begin_opt(int argc, char *const *argv, const char *getopts_string
     opts->preserve = false;
     opts->running = true;
     opts->verbose = false;
+    opts->write_seed = 0;
+    opts->read_seed = 0;
```
Comment (sueloverso) [inline test/utility/parse_opts.c:152]: Here

File: test/utility/parse_opts.c [hunk 7]
```diff
@@ -95,7 +166,8 @@ testutil_parse_begin_opt(int argc, char *const *argv, const char *getopts_string
       USAGE_STR('A', " [-A append thread count]"), USAGE_STR('b', " [-b build directory]"),
       USAGE_STR('C', " [-C]"), USAGE_STR('d', " [-d add data]"), USAGE_STR('h', " [-h home]"),
       USAGE_STR('m', " [-m]"), USAGE_STR('n', " [-n record count]"),
-      USAGE_STR('o', " [-o op count]"), USAGE_STR('P', " [-PT] [-Po storage source]"),
+      USAGE_STR('o', " [-o op count]"),
+      USAGE_STR('P', " [-PT] [-Po storage source] [-PSW<write_seed>,R<read_seed>]"),
```
Comment (sueloverso) [inline test/utility/parse_opts.c:170]: Here

File: test/utility/parse_opts.c [hunk 8]
```diff
@@ -132,8 +204,14 @@ testutil_parse_end_opt(TEST_OPTS *opts)
...
 
-    if (opts->tiered_storage && opts->tiered_storage_source == NULL)
-        opts->tiered_storage_source = dstrdup(DIR_STORE);
+    if (opts->tiered_storage) {
+        if (opts->tiered_storage_source == NULL)
+            opts->tiered_storage_source = dstrdup(DIR_STORE);
+
+        /* Initialize the state for the random number generators. */
+        parse_init_random(&opts->write_rnd, &opts->write_seed);
+        parse_init_random(&opts->read_rnd, &opts->read_seed);
```
Comment (sueloverso) [inline test/utility/parse_opts.c:213]: Here

File: test/utility/test_util.h [hunk 9]
```diff
@@ -79,6 +79,13 @@ typedef struct {
     FILE *progress_fp; /* Progress tracking file */
     char *progress_file_name;
 
+    WT_RAND_STATE write_rnd; /* PRNG state for write ops */
+    WT_RAND_STATE read_rnd;  /* PRNG state for read ops */
+    uint64_t write_seed;     /* Random seed for write ops */
+    uint64_t read_seed;      /* Random seed for read ops */
```
Comment (sueloverso) [inline test/utility/test_util.h:85]: Can you alphabetize read before write in the structure?

File: test/utility/test_util.h [hunk 10]
```diff
@@ -79,6 +79,13 @@ typedef struct {
...
+
+#define TESTUTIL_SEED_FORMAT "-PSW%" PRIu64 ",R%" PRIu64
```
Comment (sueloverso) [inline test/utility/test_util.h:87]: Here too.

File: test/csuite/wt9937_parse_opts/main.c [hunk 11]
```diff
@@ -153,6 +175,8 @@ report(const TEST_OPTS *opts, FICTIONAL_OPTS *fiction_opts)
     REPORT_STR(opts, build_dir);
     REPORT_STR(opts, tiered_storage_source);
     REPORT_INT(opts, table_type);
+    REPORT_INT(opts, write_seed);
+    REPORT_INT(opts, read_seed);
```
Comment (sueloverso) [inline test/csuite/wt9937_parse_opts/main.c:179]: Here

**Review by sueloverso — APPROVED** [review]
I have a bunch of alphabetical nits and would prefer 'read' to come before 'write'.

**Comment by ddanderson** [comment]
Hi @sueloverso, I understand the desire for alphabetizing and the natural feeling that reads precede writes.  But there's a reason to put write seeds first.  If you remember, writes will dictate the precise keys and operations so that something can be predictably replayed if the same write seed is used.  The read seed controls all the other stuff that has no bearing on what the data ends up looking like.  So if I'm running a flushing node and coordinate runs with an accepting node, I need the write seeds to match, but read seeds don't matter.  It seemed like a good idea (in terms of how it will be used copy/paste) that `"-PSW1234,R5678"` reported by the flushing could be easily copied to the accepting node as `"-PSW1234"` (where it selects its own read seed).  It's also a little easier to eyeball (and code) that `"-PSW1234,R5678"` works with `"-PSW1234,R9753".`  So that's where the ordering is coming from.

We could be call them primary seed and secondary seed, (or data seed and misc seed, or first seed and second seed) or any other terminology that makes the ordering seem more natural.  Even A seed and B seed works.

**Review by raviprakashgiri29 — APPROVED** [review]
LGTM

**Comment by sueloverso** [comment]
> We could be call them primary seed and secondary seed, (or data seed and misc seed, or first seed and second seed) or any other terminology that makes the ordering seem more natural. Even A seed and B seed works.

Thank you for the detailed explanation @ddanderson. I think the key difference for me is that, in my mind, read and write are things that go vertically/down into the system. Your description of the flushing node and coordinated accepting node(s) is a horizontal/across thing in my head. So some terminology that reflects that would seem more natural. Based on your description the data seed and misc seed make sense (or 'data' and 'other'). But primary and secondary might work well enough too.

**Comment by ddanderson** [comment]
Hi @sueloverso , what do you think of "data" seed and "extra" seed?  I like the 'd'/'e', will make it easier to remember the pair together, at least for me.

**Comment by sueloverso** [comment]
data/extra would work for me.

**Comment by ddanderson** [comment]
evergreen merge

**Pull request merged**